### DISCUSSION TITLE
I found a bug or deficiency in the code - is this the right place to mention it?

### DISCUSSION THREAD

**Discussion by Ives0721** [Mathics3/mathics-core#860] [General]
[ImageAdjust](https://reference.wolfram.com/language/ref/ImageAdjust.html) in Wolfram refreence has a rule like this:

```
ImageAdjust[ image, corr, {inmin, inmax}, {outmin, outmax} ]
```

It rescales the image so that the range of input values `inmin` to `inmax` is mapped to `outmin` to `outmax`. But it haven't implemented in the source code

**Comment by rocky** [comment]
Are you up for implementing it?

**Reply by Ives0721** [reply]
Sorry, modify the Mathics'source code would be difficult for me at my current programming skill. 😅

**Reply by mmatera** [reply]
@Ives0721, let me reformulate the question: Would you be able to write a WMA expression that does the same, but using the currently implemented subset of the WL in Mathics?

**Reply by Ives0721** [reply]
> @Ives0721, let me reformulate the question: Would you be able to write a WMA expression that does the same, but using the currently implemented subset of the WL in Mathics?

I'm sorry to say this, but I think you can regard me as a newbie researcher interested in Mathics for now, rather than a developer. So  I think i can't implement this function and contribute to Mathics. 
On the one hand, I am currently working on several papers and it is hard to find time to try. On the other hand, I'm not sure about some of the details of `ImageAdjust` (how it rescale?  i haven't try it on MMA).  And i haven't use MMA for a long time, which makes me forget its grammar 😅. So i hope you can forgive me and let my inappropriate question

**Reply by mmatera** [reply]
Regarding the time available, I think we are all in the same situation. Here is scheme of the idea about how to implement it, if you have the will:
* Use `ImageData` to convert the Image into numeric arrays.
* Apply a transformation element by element. In this case I think a linear transformation could work.
* Use `Image[...]` to convert the resulting array from the previous step into a new Image.

If you can do that, share the expression and then we could add it to a package.

**Comment by rocky** [comment]
Ok - other than the fact that Mathics3 is incomplete, is there a particular reason why this particular deficiency would benefit the community above the thousands of other deficiencies?

**Comment by rocky** [comment]
Since this was filed under "General Discussions" let me take the opportunity to broaden something that is more of a general discussion rather than a bug report or feature request. https://github.com/Mathics3/mathics-core/issues is where one files bug reports and feature requests. Issues filed there have labels to categorize the kinds of requests, they are better integrated into the Pull Request mechanism, and one can more easily track these in aggregation. 

As I mentioned above, there are a vast number of ways that Mathics3 could be improved. A rough estimate is in the hundreds if not thousands of ways.  Some of these things are fundamental to improving Mathics, and some that don't have such an impact.

At any given time, the number of people who are actively improving Mathics3 can be counted on the number of thumbs a person often has. 

Furthermore, this is all volunteer time.  And for me this isn't the only open-source project I work on. 

Given this, I find myself constantly asking myself, of the many things I could be doing, even given open source things, which things are the most impactful?  And which are the kinds of things I want to spend volunteer time on?

In issue reports, we suggest that people fill out "priority" information. If this problem is keeping you from releasing an open source project that helps the world in some way, let us know! If you are a packager for some software distribution and this prevents the package from getting released, let us know! 

I imagine most of the time though someone is using the code and finds something that is noticed. And I appreciate that some time was spent to even file a "general discussion". 

But maybe the problem isn't a big deal. Maybe it is in the category of nice to have but not necessary.   And that is why in issue reports we ask people to indicate this. 

Also, we have a section called in the  issue form called "Workarounds". For something like this, it is likely that some code in Mathics3 could be written that uses the existing functionality to do what is needed. Mathics3 and WMA are extensible in this regard.  

The most common response I get when I ask if a bug reporter or issue requester would be willing to fix the problem or suggest a workaround is a comment about their difficulty in programming or contemplating such a thing, sometimes with some cute emoji, or equivalent remark to soften the ultimate "no".  

Let me say that no one is born, knowing how to programming. It takes some effort to do so. In this project in particular, I have spent some effort in https://mathics-development-guide.readthedocs.io/en/latest/ describe how the code works precisely to make it easier for others to learn and modify the code. 

But, okay, not everyone can or is willing to do everything one might want to see done.  Pretty universally, that is why people get paid to do work. You could fund the project. Or if you don't want to do that, you could pay a programmer to fix the code, for example offer a bounty for fixing the problem on some bounty platform.

 I am even open to bartering your service for something you would like me to do.   Open source code is one of the few areas where it feels normal to make a request for something you don't want to expend time, effort, or money on but have no compunction in suggesting others work on something you'd like done.

**Reply by Ives0721** [reply]
I apologize for my inappropriate question without unsdestanding how GitHub works. And thanks for your guiding, now I'am trying to read Mathics’s developer reference in my spare time.

**Reply by rocky** [reply]
> I apologize for my inappropriate question without unsdestanding how GitHub works. And thanks for your guiding, now I'am trying to read Mathics’s developer reference in my spare time.

No apologies needed. 

This kind of thing happens somewhat naturally and is somewhat expected. But that's why it might be helpful to note this somewhere. I am sure I have reported problems many times and places in other projects without giving much thought or without considering: 

*  is this a free version of some supported project, or purely open source? (fixing bugs in free versions of product can assist in the fixing bugs in the paid versions) 
*  are there lots of contributors to the project? 
* how many bugs exist and how quickly are they fixed? 
* what are the processes for reporting a bug on this venue? For example might not be familiar with how things typically work on github?
